bill/invoice_convert.go

```diff
@@ -0,0 +1,173 @@
+package bill
+
+import (
+	"fmt"
+
+	"github.com/invopop/gobl/currency"
+	"github.com/invopop/gobl/pay"
+)
+
+// ConvertInto will use the defined exchange rates in the invoice to convert all the prices
+// into the given currency.
+//
+// The intent of this method is help convert the invoice amounts when the destination is
+// unable or unwilling to handle the current currency. This is typically the case
+// with tax related reports or declarations.
+//
+// The method will return a new invoice with all the amounts converted into the given
+// currency or an error if the conversion is not possible.
+//
+// Conversion is done by first exchanging the lowest common amounts to the destination
+// currency, then recalculating the totals.
+func (inv *Invoice) ConvertInto(cur currency.Code) (*Invoice, error) {
+	// Calculate ensures that all the totals and amounts have been prepared
+	// so we can make assumptions about the data that will be available,
+	// including the original currency!
+	if err := inv.Calculate(); err != nil {
+		return nil, err
+	}
+
+	if inv.Currency == cur {
+		return inv, nil
+	}
+	ex := currency.MatchExchangeRate(inv.ExchangeRates, inv.Currency, cur)
+	if ex == nil {
+		return nil, fmt.Errorf("no exchange rate defined for '%v' to '%v'", inv.Currency, cur)
+	}
+
+	i2 := *inv
+	i2.Totals = new(Totals)
+	i2.Lines = inv.convertLines(ex)
+	i2.Discounts = inv.convertDiscounts(ex)
+	i2.Charges = inv.convertCharges(ex)
+	i2.Outlays = inv.convertOutlays(ex)
+	i2.Payment = inv.convertPayment(ex)
+	i2.Currency = cur
+
+	if err := i2.Calculate(); err != nil {
```

Comment:
I think the approach makes sense, first convert the amounts in the invoice, and the calculate again. This will make the totals consistent with the rest of the invoice (i.e. not rounding discrepancies)

Reply:
Also, there are a large number of amounts to convert! 😆